doc: blog post on ciclops (#110)
cloudnative-pg/cloudnative-pg.github.io: content/blog/introducing-ciclops/index.md

---

> title: "Understading CI/CD test results: Introducing Ciclops"

Reviewer: ```suggestion
title: "Introducing Ciclops, a new GitHub action to better understand CI/CD test results"
```

date: 2023-01-25T13:13:36+01:00
draft: false
image:
    url: 7th_voyage_of_Sinbad_-_Cyclops.png
    attribution: from <strong><a href="https://commons.wikimedia.org/wiki/File:7th_voyage_of_Sinbad_-_Cyclops.png">Wikimedia Commons</a></strong>
author: jsilvela
tags:
 - blog
 - information
 - continuous integration
 - ci/cd

>  - testing

Reviewer: ```suggestion
 - testing
 - github action
 - devops
 - continuous delivery
```

summary: Introducing Ciclops, a GitHub Action to make sense of Testing
---
At CloudNativePG we take testing seriously; particularly end-to-end testing. The
CloudNativePG operator needs to work across different versions of Kubernetes and
PostgreSQL. We have an extensive test suite (over 100 end-to-end tests) that we
run on several combinations of versions of Postgres and Kubernetes, as part of

> our Continuous Integration pipeline.

Reviewer: ```suggestion
our Continuous Integration pipeline.

The goal of this initiative is to enhance the continuous delivery posture of our software,
by building quality directly into the product. Such tests indeed prevent that new features
developed in separate branches do not introduce any regression and, at the same time,
enrich the set of tests before we merge the commit on the *trunk*. This is what ensures
continuous delivery, and at the same time allows us to say that the latest snapshot on main
is the best version of CloudNativePG so far. Ever.
```

We are using GitHub Actions in our github repository, and a
[`strategy: matrix`](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs)
handles generating jobs for each combination of versions of Kubernetes and
Postgres we want to run the test suite on.

This can blow up quickly. In the image below, you can see a matrix with 9
*branches*, five of which have errors in the End-to-End Test suite.

![matrix](matrix-blowup.png)

Without going into each branch to investigate which tests fail, we don't know if
there is a pattern to the failures. Did the same tests fail on all the branches?
On branches with failures, are there many failed tests, or few? Are there any
noteworthy patterns to the failures?

Investigating test failures becomes a time sink. And 9 branches are not even
that much; internally at EDB we test more combinations, including the different

> cloud providers, for our closed source operators. It is not uncommon for us to

Reviewer: ```suggestion
cloud providers, for our EDB Postgres for Kubernetes operator. It is not uncommon for us to
```

but only two tests have failed in the whole suite (2 test combinations out of
12340). The failed tests affect two versions of Kubernetes and two versions of
PostgreSQL.
Right away we can start getting an idea. Perhaps, in this case, we could
conclude that the `Backup and restore` tests are showing a bit of brittleness.
We might try again with laxer timeouts in some of the test assertions. But we
could also conclude that the tests are mostly in good shape.

CIclops is available with the same open-source license and collaboration model
as CloudNativePG.
This is [the repository](https://github.com/cloudnative-pg/ciclops).
Feel free to use it, fork it, and contribute to it!

## Technical details

CIclops is a [GitHub Action](https://docs.github.com/en/actions), built with
Python and deployed using a Docker image.
The documentation contains explanations on building the GitHub Action, so it
might serve as a good tutorial if you want to build your own Docker-based GitHub
Action.

While the development of CIclops has been guided by our use case of testing a
Kubernetes operator for PostgreSQL databases, the model is usable more

> generally:

Reviewer: ```suggestion
generally to:
```

---
title: "Introducing Ciclops, a new GitHub action to better understand CI/CD test results"
date: 2023-01-25T13:13:36+01:00
draft: false
image:
    url: 7th_voyage_of_Sinbad_-_Cyclops.png
    attribution: from <strong><a href="https://commons.wikimedia.org/wiki/File:7th_voyage_of_Sinbad_-_Cyclops.png">Wikimedia Commons</a></strong>
author: jsilvela
tags:
 - blog
 - information
 - continuous integration
 - ci/cd
 - testing
 - github action
 - devops
 - continuous delivery
summary: Introducing Ciclops, a GitHub Action to make sense of Testing
---
At CloudNativePG we take testing seriously; particularly end-to-end testing. The
CloudNativePG operator needs to work across different versions of Kubernetes and
PostgreSQL. We have an extensive test suite (over 100 end-to-end tests) that we
run on several combinations of versions of Postgres and Kubernetes, as part of
our Continuous Integration pipeline.

> The goal of this initiative is to enhance the continuous delivery posture of our software,

Author: ```suggestion
The goal of all this testing is to enhance the continuous delivery posture of our software,
```

> by building quality directly into the product. Such tests indeed prevent that new features

Author: ```suggestion
by building quality directly into the product. Such tests indeed ensure that new features
```

developed in separate branches do not introduce any regression, as we need them to
pass before we merge the commit on the *trunk*. Moreover, each new patch should
enrich the tests of sets.

> This is what ensures continuous delivery, and at the same time allows us to say that

Author: ```suggestion
This is what enables continuous delivery, and at the same time allows us to say that
```
